iree/tools/android/run_module_app/build_apk.sh

```diff
@@ -0,0 +1,267 @@
+#!/bin/bash
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+#########################
+# Script input parameters
+#########################
+
+echo "=== A script for building iree-run-module apk ==="
+echo "This script wraps iree-run-module together with a"
+echo "specific IREE VM module and invocation information."
+echo ""
+echo "At a minimum, it expects the following env vars:"
+echo "* ANDROID_SDK_ROOT"
+echo "* ANDROID_NDK_ROOT"
+echo "* ANDROID_SDK_BUILD_TOOLS_VERSION"
+echo "See the script for details and more controls."
+echo "===-------------------------------------------==="
+echo ""
+
+print_usage_and_exit() {
+  echo "Usage: $0"
+  echo "       -d <driver>"
+  echo "       -m <input-module-file> "
+  echo "       -e <entry-function> "
+  echo "       -i <input-buffer-file> "
+  echo "       <artifact-directory>"
+  exit 1
+}
+
+# Parse module invocation options.
+while getopts ':d:m:e:i:' OPTION; do
+  case "${OPTION}" in
+    d)
+      IREE_DRIVER="${OPTARG}"
+      ;;
+    m)
+      IREE_INPUT_MODULE_FILE=$(readlink -f "${OPTARG}")
+      ;;
+    e)
+      IREE_ENTRY_FUNCTION="${OPTARG}"
+      ;;
+    i)
+      IREE_INPUT_BUFFER_FILE=$(readlink -f "${OPTARG}")
+      ;;
+    ?) # Unsupported flags
+      print_usage_and_exit
+      ;;
+  esac
+done
+shift $((OPTIND -1))
+
+# Parse artifact directory.
+if [[ $# == 1 ]]; then
+  IREE_ARTIFACT_ROOT=$(readlink -f $1)
+fi
+
+if [[ -z "${IREE_ARTIFACT_ROOT+x}" ]] || [[ -z "${IREE_INPUT_MODULE_FILE+x}" ]] || \
+   [[ -z "${IREE_ENTRY_FUNCTION+x}" ]] || [[ -z "${IREE_INPUT_BUFFER_FILE+x}" ]] || \
+   [[ -z "${IREE_DRIVER+x}" ]]; then
```

Comment:
Ah you are right. That's a subtle point I missed. The doc for it is "If parameter is null or unset, nothing is substituted, otherwise the expansion of word is substituted." I thought "null" here means empty..